File: libs/power-log-parser/src/lib/parsers/turn-cleanup-parser.ts
import { GameTag, Step } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventProvider } from '../game-event';
import { Node, TagChange } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class TurnCleanupParser implements ActionParser {
	readonly ParserName = 'TurnCleanupParser';

	private GameState: GameState;
	private ParserState: ParserState;

	constructor(parserState: ParserState, _stateFacade: StateFacade) {
		this.ParserState = parserState;
		this.GameState = parserState.GameState;
	}

	AppliesOnNewNode(node: Node, stateType: StateType): boolean {
		if (stateType !== StateType.GameState || node.Type !== TagChange) return false;
		const tc = node.Object as TagChange;
		const isNormalTurnChange =
			!this.ParserState.IsMercenaries() &&
			tc.Name === (GameTag.TURN as number) &&
			this.GameState.GetGameEntity()?.Entity === tc.Entity;
		const isMercenariesTurnChange =
			this.ParserState.IsMercenaries() &&
			tc.Name === (GameTag.STEP as number) &&
			tc.Value === (Step.MAIN_PRE_ACTION as number) &&
			this.GameState.GetGameEntity()?.Entity === tc.Entity;
		return isNormalTurnChange || isMercenariesTurnChange;
	}

	AppliesOnCloseNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		const tc = node.Object as TagChange;
		const newTurnValue = tc.Name === (GameTag.TURN as number) ? tc.Value : this.GameState.CurrentTurn + 1;
		this.GameState.CurrentTurn = newTurnValue;
		this.GameState.OnNewTurn();
		return null;
	}

	CreateGameEventProviderFromClose(_node: Node): GameEventProvider[] | null {
		return null;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/TurnCleanupParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;
using System.Linq;

namespace HearthstoneReplays.Events.Parsers
{
    // This class is just to be able to call GameState.OnNewTurn()
    public class TurnCleanupParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public TurnCleanupParser(ParserState ParserState, StateFacade helper)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = helper;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            var isNormalTurnChange = !ParserState.IsMercenaries() 
                && node.Type == typeof(TagChange)
                && (node.Object as TagChange).Name == (int)GameTag.TURN
                && GameState.GetGameEntity()?.Entity == (node.Object as TagChange).Entity;
            // While the TURN tag is present in mercenaries, it is incremented on the Innkeeper entity,
            // and the logs don't let us easily disambiguate between the AI Innkeeper and the player's
            // Innkeeper, so we rely on the turn structure tags instead
            var isMercenariesTurnChange = ParserState.IsMercenaries() 
                && node.Type == typeof(TagChange)
                && (node.Object as TagChange).Name == (int)GameTag.STEP
                && (node.Object as TagChange).Value == (int)Step.MAIN_PRE_ACTION
                && GameState.GetGameEntity()?.Entity == (node.Object as TagChange).Entity;
            return stateType == StateType.GameState
                && (isNormalTurnChange || isMercenariesTurnChange);
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return false;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            var tagChange = node.Object as TagChange;
            var newTurnValue = tagChange.Name == (int)GameTag.TURN ? (int)tagChange.Value : GameState.CurrentTurn + 1;
            GameState.CurrentTurn = newTurnValue;
            GameState.OnNewTurn();
            return null;
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            return null;
        }
    }
}
